collectors/python.d.plugin/alarms/alarms.chart.py

```diff
@@ -51,21 +59,20 @@ def _get_data(self):
         alarms = raw_data.get('alarms', {})
 
         data = {a: self.sm[alarms[a]['status']] for a in alarms if alarms[a]['status'] in self.sm}
-        self.update_charts(alarms, data)
+        self.update_charts('alarms', data)
         data['alarms_num'] = len(data)
 
+        if self.show_alarm_values:
+            data_values = {'{}_value'.format(a): alarms[a]['value'] * 100 for a in alarms if 'value' in alarms[a] and alarms[a]['value'] is not None}
+            self.update_charts('values', data_values, divisor=100)
+            data = {**data, **data_values}
+
         return data
 
-    def update_charts(self, alarms, data):
+    def update_charts(self, chart, data, algorithm='absolute', multiplier=1, divisor=1):
         if not self.charts:
             return
 
-        for a in data:
-            if a not in self.collected_alarms:
-                self.collected_alarms.add(a)
-                self.charts['alarms'].add_dimension([a, a, 'absolute', '1', '1'])
-
-        for a in list(self.collected_alarms):
-            if a not in alarms:
-                self.collected_alarms.remove(a)
-                self.charts['alarms'].del_dimension(a, hide=False)
```

Comment:
Don't you need to delete dimensions in the `alarms` chart anymore?

Reply:
I don't think so - if it just disappears from the `data` from the collector then i think it will disappear from the chart. The trouble i had was trying to manage the bookkeeping for `self.collected_alarms` in the situation where we are charting both `alarms` and `values` chart. It was confusing for me to consider all the combinations that could happen if we had a status for one alarm and not a value or vice versa. In such a case, removing the alarm from `self.collected_alarms` was breaking the other chart (and vice versa). 

So i was trying to just avoid having to track and manage `self.collected_alarms` as was not quite sure why we need it. 

I think by default such dimensions will be hidden. 

Or maybe i could just use both a `self.collected_alarms` and `self.collected_values` and then bring back this code and do the bookkeeping based on if trying to update the alarms or the values chart - kinda separate it out.

Reply:
> as was not quite sure why we need it.

We need it because `charts.dimensions` is a `list` and it is O(n) to check `if dim in charts['id']`. If there are a lot of dimensions it is slow. So there are 2 ways:

 - refactor charts dimensions attr (make it ordered set or somthing)
 - use local cache for checks (self.collected_alarms)

---


> The trouble i had was trying to manage the bookkeeping for self.collected_alarms in the situation where we are charting both alarms and values chart.

Just add a second one

 - self.collected_status_alarms
 - self.collected_values_alarms

Reply:
> I think by default such dimensions will be hidden.

No, there is auto-cleanup for (stale/not updated) charts and there is no such things for dimensions. Have to be removed manually via `self.charts['id'].dim_remove`

Reply:
@andrewm4894 ofc i would prefer _refactor charts dimensions_, but let's keep using local cache until we have done it

Reply:
Yep, agree just adding second one is the best option here, will do

Reply:
i have added dict of sets called `self.collected_dims` to keep track of the set of collected dims for each chart and maintain it in the same way as the current alarms collector does.